### PR TITLE
TRAIN-266: Add login page.

### PR DESCRIPTION
**Pull request by r-barlow**

### CODE REVIEW

File: src/container/Login.container.jsx [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+                setError("Unknown server error!");
+            }
+            return;
+        }
+
+        const token = response.data['token'];
+        if (token === undefined) {
+            setError("Unknown server error!");
+            return;
+        }
```
Comment (colsen-neslo) [inline src/container/Login.container.jsx]: I am not sure if you have looked at any training material with regards to the redux-saga pattern? So we like to follow this pattern whereby we get our containers to dispatch actions that sagas will watch for. The saga will then perform http requests with axios and then we use redux state management to help manage application state.

Please read up on some of that and you can come back and modify the current implementation to use the redux-saga pattern in a separate task.

File: src/container/Login.container.jsx [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+    const onUsernameChange = useCallback((value) => {
+        setError('');
+        setUsername({current: value, error: false});
+    }, []);
+
+    const onPasswordChange = useCallback((value) => {
+        setError('');
+        setPassword({current: value, error: false});
+    }, []);
```
Comment (colsen-neslo) [inline src/container/Login.container.jsx]: So another way to further improve this is if you have a form state object that holds the username and password values as fields, you can then create a single on change function that can consume the field name and the value and you can perform a single on change like something below, where the `newState` value represents something like `{ username: 'exampleusername' }` 
```suggestion
    const onChange = (newState) => {
        setForm({ ...form, ...newState });
    };
```

File: src/container/Login.container.jsx [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    return (
+        <>
```
Comment (colsen-neslo) [inline src/container/Login.container.jsx]: You can remove the react fragment if you are only returning a single element.

File: src/page/Login.jsx [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import Input from "../component/form/Input";
+
+import './login.scss';
+
+const Login = ({handleSubmit, onUsernameChange, onPasswordChange, username, password, error}) => {
+
+    return (<>
+        <div className="login__background"></div>
+        <div className="login__card__frame">
+            <div className="login__card">
```
Comment (colsen-neslo) [inline src/page/Login.jsx]: The card can be defined as an article element.
```suggestion
            <article className="login__card">
```

File: src/page/Login.jsx [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+import Input from "../component/form/Input";
+
+import './login.scss';
+
+const Login = ({handleSubmit, onUsernameChange, onPasswordChange, username, password, error}) => {
+
+    return (<>
+        <div className="login__background"></div>
+        <div className="login__card__frame">
```
Comment (colsen-neslo) [inline src/page/Login.jsx]: You can make this a section.
```suggestion
        <section className="login__card__frame">
```

File: src/page/Login.jsx [hunk 6]
```diff
@@ -0,0 +1,45 @@
+import PropTypes from "prop-types";
+
+import Input from "../component/form/Input";
+
+import './login.scss';
+
+const Login = ({handleSubmit, onUsernameChange, onPasswordChange, username, password, error}) => {
+
+    return (<>
+        <div className="login__background"></div>
```
Comment (colsen-neslo) [inline src/page/Login.jsx]: You can self close an empty tag.
```suggestion
        <div className="login__background" />
```

File: src/stylesheet/00_settings/_dimen.scss [hunk 7]
```diff
@@ -0,0 +1,81 @@
+$login: (
```
Comment (colsen-neslo) [inline src/stylesheet/00_settings/_dimen.scss]: What is the purpose of this file? Does it just define dimension variables for certain components?
If so why not just define them in the component stylesheets as these should never be accessed out the scope of the login component?

File: src/tool/auth.util.jsx [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+            'Content-Type': 'application/json',
+            ...getAuthHeaders()
+        },
+        body: JSON.stringify(user)
+    }).then(response => {
+        return response.json().then(result => {
+            return {status: response.status, success: response.ok, data: result};
+        });
+    }).catch(e => console.log(e))
+}
```
Comment (colsen-neslo) [inline src/tool/auth.util.jsx:48]: As mentioned earlier, once you implement redux-saga patterns this http call will be replaced by the saga that contains the axios http call to the API.

**Review by colsen-neslo — CHANGES_REQUESTED** [review]

File: src/component/form/input.scss [hunk 9]
```diff
@@ -0,0 +1,84 @@
+@use "../../stylesheet/00_settings/color" as *;
+@use "../../stylesheet/00_settings/util" as *;
+
+$dimen: (
```
Comment (colsen-neslo) [inline src/component/form/input.scss]: Is this necessary, can you not define it inline below where required?

File: src/component/ui/tooltip.scss [hunk 10]
```diff
@@ -0,0 +1,103 @@
+@use "../../stylesheet/00_settings/color" as *;
+@use "../../stylesheet/00_settings/util" as *;
+
+$dimen: (
```
Comment (colsen-neslo) [inline src/component/ui/tooltip.scss:4]: Same as above?

File: src/page/Login.jsx [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+import './login.scss';
+
+const Login = ({handleSubmit, onChange, form}) => {
+
+    return (<>
+        <div className="login__background"/>
+        <section className="login__card__frame">
+            <article className="login__card">
+                <div className="login__card__body">
```
Comment (colsen-neslo) [inline src/page/Login.jsx]: Is this div necessary? I think we can remove this if there's no specific reason to have it.

File: src/page/login.scss [hunk 12]
```diff
@@ -0,0 +1,166 @@
+@use "../stylesheet/00_settings/color" as *;
+@use "../stylesheet/00_settings/util" as *;
+@use "../stylesheet/00_settings/extra" as *;
+@use "../stylesheet/01_tools/mixin" as mixins;
+@use "../stylesheet/01_tools/const" as const;
+
+$dimen: (
```
Comment (colsen-neslo) [inline src/page/login.scss]: Same question as previous styles?

**Review by colsen-neslo — CHANGES_REQUESTED** [review]

**Review by colsen-neslo — APPROVED** [review]

File: src/component/form/Input.js [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+const Input = ({type, placeholder, hasError, onChange}) => {
+
+    const inputRequiredError = () => {
+        return (
+            <div className="form__input-error__image__frame">
+                <img className="form__input-error__image" src="/images/error.svg" alt="input error"/>
+                <Tooltip text="Required" position={POSITION.RIGHT}/>
+            </div>
+        );
+    }
```
Comment (JOlsen-Neslo) [inline src/component/form/Input.js]: We tend to refrain from following this sort of pattern, namely `Indirection`, as this doesn't make it clear what the flow of the markup is going to follow. Creating functions that render JSX generally means we should rather be creating a private component to render this markup

File: src/component/form/Input.js [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+
+import Tooltip from "../ui/Tooltip";
+
+import "./input.scss";
+
+const Input = ({type, placeholder, hasError, onChange}) => {
+
+    const inputRequiredError = () => {
+        return (
+            <div className="form__input-error__image__frame">
```
Comment (JOlsen-Neslo) [inline src/component/form/Input.js]: If you are following the BEM structure for your styling, this breaks those rules. It should always be `block__element--modifer`

What this looks like is `block_element_element_element`

File: src/component/form/Input.js [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+                <img className="form__input-error__image" src="/images/error.svg" alt="input error"/>
```
Comment (JOlsen-Neslo) [inline src/component/form/Input.js]: Rather import your SVG as a component

`import { ReactComponent as ErrorIcon } from '../images/error.svg';`

You can then render it like this:

`<ErrorIcon />`

**Review by JOlsen-Neslo — CHANGES_REQUESTED** [review]

File: src/component/form/Input.jsx [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+                   className={`form__control${hasError ? " form__input__error" : ""}`}
+                   placeholder={placeholder}
+                   aria-label={placeholder}
+                   onChange={e => {
+                       onChange(e.target.value);
+                   }}/>
+            {hasError ? <InputRequired/> : null}
+        </div>
+    );
+};
```
Comment (JOlsen-Neslo) [inline src/component/form/Input.jsx]: When there is no JavaScript in the component, no need for curly braces

```suggestion
const Input = ({type, placeholder, hasError, onChange}) => 
     <div className="form__input__group">
          <input type={type}
                 className={`form__control${hasError ? " form__input__error" : ""}`}
                 placeholder={placeholder}
                 aria-label={placeholder}
                 onChange={e => {
                     onChange(e.target.value);
                 }}/>
          {hasError ? <InputRequired/> : null}
      </div>
);
```

File: src/component/form/InputRequired.jsx [hunk 17]
```diff
@@ -0,0 +1,18 @@
+import {POSITION} from "../../tool/constant";
+
+import Tooltip from "../ui/Tooltip";
+
+import { ReactComponent as ErrorIcon } from '../../icon/error.svg';
+
+import "./input-required.scss";
+
+const InputRequired = () => {
```
Comment (JOlsen-Neslo) [inline src/component/form/InputRequired.jsx]: Same comment here

**Review by JOlsen-Neslo — CHANGES_REQUESTED** [review]

File: src/component/form/InputRequired.jsx [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+
+import Tooltip from "../ui/Tooltip";
+
+import {ReactComponent as ErrorIcon} from '../../icon/error.svg';
+
+import "./input-required.scss";
+
+const InputRequired = () => (
+    <div className="form__input__error__image__frame">
+        <ErrorIcon className="form__input__error__image" alt="input error"/>
```
Comment (JOlsen-Neslo) [inline src/component/form/InputRequired.jsx]: These classes both don't follow the BEM structure

File: src/page/Login.jsx [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+
+import {ReactComponent as Logo} from "../icon/insight.svg";
+
+import './login.scss';
+
+const Login = ({handleSubmit, onChange, form}) => {
+
+    return (<>
+        <div className="login__background"/>
+        <section className="login__card__frame">
```
Comment (JOlsen-Neslo) [inline src/page/Login.jsx]: a few classes in this file also don't follow BEM

**Review by JOlsen-Neslo — CHANGES_REQUESTED** [review]

**Review by JOlsen-Neslo — APPROVED** [review]

**Pull request merged**